### PR TITLE
Use MyST-NB to render jupyter notebook and markdown files

### PR DESCRIPTION
**Pull request by weiji14**
Render the [icepyx example jupyter notebooks](https://github.com/icesat2py/icepyx/tree/development/examples) directly at https://icepyx.readthedocs.io/en/development (i.e. without having to go to GitHub)!

**Preview** at https://icepyx--196.org.readthedocs.build/en/196/getting_started/examples.html

image

This is done using [MyST-NB](https://myst-nb.readthedocs.io/en/v0.12.0) which parses jupyter notebooks (or markdown) in Sphinx. MyST-NB also acts as a replacement for:

- nbsphinx (for .ipynb files) added in #28 at aa75a5a9a76bce66bd9ed28777a8ce667fb39253
- recommonmark (for .md files) added in #49 at 1d7919603fe641860ec80b0317a01d48ebd6c41b

The powerful [MyST (Markedly Structured Text)](https://myst-parser.readthedocs.io/en/v0.13.5) parser also allows much of the documentation to be written in Markdown instead of reStructuredText, which can make it easier for new folks wanting to make contributions to the documentation.

~~This PR also replaces [`numpydoc`](https://numpydoc.readthedocs.io/en/latest/index.html) with Sphinx's built-in [`Napolean`](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/) extension (they practically do the same thing).~~ Edit: sticking with numpydoc for now as per https://github.com/icesat2py/icepyx/pull/196#issuecomment-932335905

**P.S.** Context for this Pull Request is to eventually revamp the front-facing docs page as per #110. This is really just a maintenance PR laying the groundwork by modernizing some of the Sphinx documentation dependencies, and fixing some issues along the way.

### CODE REVIEW

**Comment by JessicaS11** [comment]
So far this looks like a great upgrade to our docs. Thanks @weiji14! I seem to remember, vaguely, some conversation around the pros/cons of using numpydoc instead of Napoleon, but I'll have to see if I can find any notes on that (@norlandrhagen, do you remember?).

My only major comment right now is that after a lot of back and forth we intentionally put the examples directory in the top level directory to make them easy to find. It's not intuitive for people to look in doc/source for them, even thought that's a "standard" home for them. I like the idea of them being rendered directly in the documentation, but also easy for people to find (since many find the repo before the docs). Thus, we should do some brainstorming around where they are available in the documentation (under "examples" or under "user guide"), where they reside in the repo, and what components of each notebook (if any besides the title) we want to show up in the left-hand menu bar, as well as how this ultimately relates to a user-contributed example gallery.

**Comment by weiji14** [comment]
> My only major comment right now is that after a lot of back and forth we intentionally put the examples directory in the top level directory to make them easy to find. It's not intuitive for people to look in doc/source for them, even thought that's a "standard" home for them. I like the idea of them being rendered directly in the documentation, but also easy for people to find (since many find the repo before the docs).

Yes, I realize there was a lot of back and forth of moving the example notebooks, and I did try really hard to keep the *.ipynb files where they were in `examples/` for discoverability. But it doesn't seem to work for MyST-NB, since the library only renders jupyter notebooks under the `doc/` folder where sphinx is running. Maybe I'm missing some special configuration, but putting it in `doc/user_guide` was the easiest way I could get the notebook to show up in the docs.

> Thus, we should do some brainstorming around where they are available in the documentation (under "examples" or under "user guide"), where they reside in the repo, and what components of each notebook (if any besides the title) we want to show up in the left-hand menu bar, as well as how this ultimately relates to a user-contributed example gallery.

Yes, an example gallery (as per #27) would be cool! I was catching up on the stalled issue at https://github.com/icesat2py/icepyx-gallery/issues/1 the other day too. Maybe something to discuss for the meeting tomorrow (though I won't make it at 4am) or later this month.

**Comment by weiji14** [comment]
I'll try and put some work into this PR (and possibly #110) this week.

> So far this looks like a great upgrade to our docs. Thanks @weiji14! I seem to remember, vaguely, some conversation around the pros/cons of using numpydoc instead of Napoleon, but I'll have to see if I can find any notes on that (@norlandrhagen, do you remember?).

There's a stackoverflow question on it at https://stackoverflow.com/questions/36715801/difference-between-sphinxcontrib-napoleon-and-numpy-numpydoc#45009645. Just to quote it:

> default behavior of napoleon links to known datatypes in the python documentation, and it is slightly more condensed (numpydoc displays a bit like how it appears in the docstring)

Functionally, they seem to do about the same thing (save us from writing in raw Sphinx syntax). I can revert the Napolean addition and keep things on NumpyDoc if you prefer to keep things as it is?

> My only major comment right now is that after a lot of back and forth we intentionally put the examples directory in the top level directory to make them easy to find. It's not intuitive for people to look in doc/source for them, even thought that's a "standard" home for them. I like the idea of them being rendered directly in the documentation, but also easy for people to find (since many find the repo before the docs). Thus, we should do some brainstorming around where they are available in the documentation (under "examples" or under "user guide"), where they reside in the repo, and what components of each notebook (if any besides the title) we want to show up in the left-hand menu bar, as well as how this ultimately relates to a user-contributed example gallery.

Another idea would be to keep the `examples/` folder for visibility, but put a link in the `examples/README.md` to point to `doc/source/example_notebooks/`, similar to how `geopandas` does it at https://github.com/geopandas/geopandas/tree/master/examples.

**Comment by JessicaS11** [comment]
> Another idea would be to keep the `examples/` folder for visibility, but put a link in the `examples/README.md` to point to `doc/source/example_notebooks/`, similar to how `geopandas` does it at https://github.com/geopandas/geopandas/tree/master/examples.

It's really weird to me that you can't use links to get to files outside of docs. It looks like it could be in the works, as [this feature](https://myst-parser.readthedocs.io/en/latest/sphinx/use.html#include-a-file-from-outside-the-docs-folder-like-readme-md) has been implemented for Markdown-style images and links. I played around with using `{include}` as in the example, but then it was just printing the command on the page rather than executing it. Next week I'll work on moving the rest of the examples and putting them all under a single examples tab with good references as you suggested r.e. `geopandas`.

File: doc/source/contributing/code_of_conduct_link.md [hunk 1]
```diff
@@ -0,0 +1,2 @@
+```{include} ../../../code_of_conduct.md
+```
```
Comment (JessicaS11) [inline doc/source/contributing/code_of_conduct_link.md:2]: What was the reasoning for switching this from an rst to a md file? Does it render differently/better?

**Comment by fperez** [comment]
I wanted to ping @choldgraf on this one :) Chris, do you have any thoughts on the examples management issue discussed above by @weiji14 and @JessicaS11? It seems that having flexibility on their location, for the purposes of regular use of the repo, is a reasonable use pattern, but it seems to clash with the rest of the build process a bit...

A rather brutish hack would be to add a recursive copy operation against he examples dir at the start of the build (we could do it with hardlinks to go fast - `os.link` [from the stdlib](https://docs.python.org/3/library/os.html#files-and-directories) works on Unix and Windows, so a quick recursive copy can be made with a single `shutil.copytree(src, dst, copy_function=os.link)` call.

But that feels crude, so perhaps @choldgraf has better ideas ;)

**Comment by choldgraf** [comment]
Hmmm one thing you could try is saving your notebooks as MyST markdown notebooks and then using an include directive in your docs, but that may not be ideal if you want them as ipynb.

To be honest i think the most reasonable solution is to indeed add a copytree in the conf.py so that it gets moved over at build time, and then add that folder to your gitignore so it isn't in the repository more than once. But annoyingly Sphinx has no way to reference file outside the documentation root from the toctree

**Comment by JessicaS11** [comment]
@weiji14 When you get a chance (no rush - January is fine!) can you look this PR over again since I was the last one to make a bunch of edits? Then I can give it an approving review and we can merge it. Note that I created Issue #257 to improve the notebook organization after this initial PR to get the functionality there.

**Comment by weiji14** [comment]
> @weiji14 When you get a chance (no rush - January is fine!) can you look this PR over again since I was the last one to make a bunch of edits? Then I can give it an approving review and we can merge it. Note that I created Issue #257 to improve the notebook organization after this initial PR to get the functionality there.

Just from a quick scan, I do feel like there's too many nested headings. Do you think it's possible to remove the 'Example notebooks' level, and just have everything under 'Examples'?

image

I'll try and give this a closer look next week after AGU is over. Too many things on my plate right now...

**Comment by JessicaS11** [comment]
> Do you think it's possible to remove the 'Example notebooks' level, and just have everything under 'Examples'?

Done - great catch.

> I'll try and give this a closer look next week after AGU is over. Too many things on my plate right now...

No rush - I just wanted to document it was ready for the next step.

**Comment by weiji14** [comment]
Hi @JessicaS11, I've made a few more minor updates to fix some Sphinx doc build warnings. Also, I've moved the notebook examples into a dedicated section on the sidebar to improve visibility:

image

Let me know if you're ok with this, otherwise I can revert commit a161b428ea2c1ccabc848ad16d3f7cd743771b44 to have the examples be nested under Getting Started -> Examples as before:

image

**Comment by JessicaS11** [comment]
> Also, I've moved the notebook examples into a dedicated section on the sidebar to improve visibility:

I really like this change, but I'm concerned that it will make the menu bar long and diluted as we add new examples. To give us time to come up with an plan (perhaps we can have a few main categories of example type that get listed on the menu bar, with individual examples listed within?), I reverted the change in this PR and after we merge will create a new branch with it that we can keep working on. That will also have your commit to increase the heading levels, since we need to do a go-through of the examples to standardize heading levels anyway.

**Review by JessicaS11 — APPROVED** [review]

**Pull request merged**